### PR TITLE
[MAINTENANCE] begin static type-checking

### PR DESCRIPTION
**Pull request by Kilo59**
Changes proposed in this pull request:
- add `type-check` invoke task that runs mypy against a subset of ge sub-packages
  - ~data_asset~
  - ~exceptions~
  - ~jupyter_ux~
  - [x] profile
  - ~self_check~
  - ~types~
- [x] Get Azure pipeline `type-check` step working

-----

### Edit
Updated PR so that it's easier to review. Only type checking the `profile` package in this PR.
Will open a separate PR with the additionally completed module typings (`data_asset`, `exceptions`, `jupyter_ux`, `self_check`, `type`).

----

[I could have used `cast` more heavily](https://adamj.eu/tech/2021/07/06/python-type-hints-how-to-use-typing-cast/) but IMO that is better used when we understand the code well (which I do not).

The most important thing this PR does is begin the process of static type-checking, we can gradually add more and more packages/modules to the list checked by `invoke type-check`.

~We may want to wait until https://github.com/great-expectations/great_expectations/pull/5525 is merged.~ Avoiding `pyproject.toml` to get this merged sooner.

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- ~I have run any local integration tests and made sure that nothing is broken.~

### CODE REVIEW

File: great_expectations/jupyter_ux/expectation_explorer.py [hunk 1]
```diff
@@ -243,9 +244,9 @@ def initialize_data_asset_state(self, data_asset) -> None:
     def set_expectation_state(self, data_asset, expectation_state, column=None) -> None:
         data_asset_name = data_asset.data_asset_name
         expectation_type = expectation_state.get("expectation_type")
-        data_asset_state = self.state["data_assets"].get(data_asset_name)
+        data_asset_state: Dict = self.state["data_assets"][data_asset_name]
 
-        data_asset_expectations = data_asset_state.get("expectations")
+        data_asset_expectations = data_asset_state.get("expectations", {})
```
Comment (Kilo59) [inline great_expectations/jupyter_ux/expectation_explorer.py:249]: These changes and the similar ones below assume that we aren't use an AttributeError exception for control flow.

File: great_expectations/data_asset/data_asset.py [hunk 2]
```diff
@@ -343,7 +343,9 @@ def wrapper(self, *args, **kwargs):
         return outer_wrapper
 
     def _initialize_expectations(
-        self, expectation_suite=None, expectation_suite_name=None
+        self,
+        expectation_suite: Union[Dict, ExpectationSuite, None] = None,
```
Comment (Kilo59) [inline great_expectations/data_asset/data_asset.py:347]: Could this also be a list?

File: great_expectations/jupyter_ux/__init__.py [hunk 3]
```diff
@@ -60,13 +60,12 @@ def set_data_source(context, data_source_type=None):
...
 <p>
 Found more than one data source in the great_expectations.yml of your project:
-<b>{1:s}</b>
+<b>{:s}</b>
 </p>
 <p>
 Uncomment the next cell and set data_source_name to one of these names.
 </p>
 """.format(
-                        data_source_type,
```
Comment (Kilo59) [inline great_expectations/jupyter_ux/__init__.py:69]: This shouldn't change anything because it wasn't being used.

File: great_expectations/profile/base.py [hunk 4]
```diff
@@ -144,20 +144,23 @@ class Profiler(metaclass=abc.ABCMeta):
...
 
-    def __init__(self, configuration: dict = None) -> None:
+    def __init__(self, configuration: Optional[dict] = None) -> None:
         self.configuration = configuration
 
-    def validate(self, item_to_validate: Any) -> None:
-        pass
+    def validate(self, item_to_validate: Any) -> bool:
+        # TODO: Re-evaluate on this behavior
+        return False
```
Comment (Kilo59) [inline great_expectations/profile/base.py:152]: I'm going to revert this change and just ignore the related type-error along with a comment.

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+[build-sytem]
+requires = ["setuptools", "wheel"]
+build-backend = "setuptools.build_meta"
+
```
[the remark this answers is not in the document]
Comment (Kilo59) [reply pyproject.toml]: Just going with `mypy.ini` for now.

File: great_expectations/profile/json_schema_profiler.py [hunk 6]
```diff
@@ -46,10 +46,10 @@ class JsonSchemaProfiler(Profiler):
         JsonSchemaTypes.BOOLEAN.value: ProfilerTypeMapping.BOOLEAN_TYPE_NAMES,
     }
 
-    def __init__(self, configuration: dict = None) -> None:
+    def __init__(self, configuration: Optional[dict] = None) -> None:
         super().__init__(configuration)
 
-    def validate(self, schema: dict) -> bool:
+    def validate(self, schema: dict) -> bool:  # type: ignore
```
Comment (Kilo59) [inline great_expectations/profile/json_schema_profiler.py]: ☝️ this return type is inconsistent with the ABC

**Review by anthonyburdi — APPROVED** [review]
LGTM! Excited to extend this further.

**Review by NathanFarmer — APPROVED** [review]
Looks good to me!

File: great_expectations/profile/json_schema_profiler.py (hunk 6 above)
Comment (cdkini) [inline great_expectations/profile/json_schema_profiler.py]: Could we use `type: ignore[specific-mypy-error-code]` and a comment why we're excluding?

File: great_expectations/profile/json_schema_profiler.py [hunk 7]
```diff
@@ -218,7 +218,7 @@ def _create_range_expectation(
     ) -> Optional[ExpectationConfiguration]:
         """https://json-schema.org/understanding-json-schema/reference/numeric.html#range"""
         object_types = self._get_object_types(details=details)
-        object_types = filter(
+        object_types = filter(  # type: ignore
```
Comment (cdkini) [inline great_expectations/profile/json_schema_profiler.py]: Same note for all ignore comments (if not too big of a lift)

File: great_expectations/profile/json_schema_profiler.py (hunk 6 above)
Comment (Kilo59) [reply great_expectations/profile/json_schema_profiler.py]: Yes, will do!
I didn't even know `mypy` supported this.
https://mypy.readthedocs.io/en/stable/error_codes.html#silencing-errors-based-on-error-codes

Comment (Kilo59) [reply great_expectations/profile/json_schema_profiler.py]: @cdkini 
Looks like we can even get `mypy` to enforce this.
https://mypy.readthedocs.io/en/stable/error_code_list2.html#check-that-type-ignore-include-an-error-code-ignore-without-code

But for some reason, I couldn't get it to work. I will revisit enforcement of this on the next typing PR.
I did update all the current `# type: ignore` comments.
